node/src/actors/chain_manager/mod.rs

```diff
@@ -341,6 +353,167 @@ fn update_transaction_pool(transactions_pool: &mut TransactionsPool, transaction
+                Output::Reveal(_) => {}
+                Output::Tally(_) => {}
+            }
+        }
+    }
+}
+
+// TODO: move to transaction_factory.rs
+/// Select enough UTXOs to sum up to `amount`.
+///
+/// On success, return a list of output pointers and their sum.
+/// On error, return the total sum of the output pointers in `own_utxos`.
+pub fn take_enough_utxos<S: std::hash::BuildHasher>(
+    own_utxos: &HashSet<OutputPointer, S>,
+    all_utxos: &UnspentOutputsPool,
+    amount: u64,
+) -> Result<(Vec<OutputPointer>, u64), u64> {
+    let mut acc = 0;
+    let mut list = vec![];
+
+    for op in own_utxos {
+        acc += all_utxos[op].value();
+        list.push(op.clone());
+        if acc >= amount {
+            break;
+        }
+    }
+
+    if acc >= amount {
+        Ok((list, acc))
+    } else {
+        Err(acc)
+    }
+}
+
+/// Build value transfer transaction with the given outputs and fee.
+pub fn build_vtt<S: std::hash::BuildHasher>(
+    mut outputs: Vec<ValueTransferOutput>,
+    fee: u64,
+    own_utxos: &HashSet<OutputPointer, S>,
+    own_pkh: PublicKeyHash,
+    all_utxos: &UnspentOutputsPool,
+) -> Result<TransactionBody, u64> {
+    let output_value: u64 = outputs.iter().map(|x| x.value).sum();
+    match take_enough_utxos(own_utxos, all_utxos, output_value + fee) {
+        Err(sum_of_own_utxos) => Err(sum_of_own_utxos),
+        Ok((output_pointers, input_value)) => {
+            let change_amount = input_value - output_value - fee;
+            if change_amount > 0 {
+                // Create change output
+                outputs.push(ValueTransferOutput {
+                    pkh: own_pkh,
+                    value: change_amount,
+                });
+            }
+
+            let inputs = output_pointers
+                .into_iter()
+                .map(|x| {
+                    Input::ValueTransfer(ValueTransferInput {
+                        transaction_id: x.transaction_id,
+                        output_index: x.output_index,
+                    })
+                })
+                .collect();
+            let outputs = outputs.into_iter().map(Output::ValueTransfer).collect();
+            let tx_body = TransactionBody::new(0, inputs, outputs);
+            Ok(tx_body)
+        }
+    }
+}
+
+/// Build data request transaction with the given outputs and fee.
+pub fn build_drt<S: std::hash::BuildHasher>(
+    dr: DataRequestOutput,
+    fee: u64,
+    own_utxos: &HashSet<OutputPointer, S>,
+    own_pkh: PublicKeyHash,
+    all_utxos: &UnspentOutputsPool,
+) -> Result<TransactionBody, u64> {
+    let dro = Output::DataRequest(dr);
+    let output_value: u64 = dro.value();
+    match take_enough_utxos(own_utxos, all_utxos, output_value + fee) {
+        Err(sum_of_own_utxos) => Err(sum_of_own_utxos),
+        Ok((output_pointers, input_value)) => {
+            let mut outputs = vec![];
+            let change_amount = input_value - output_value - fee;
```

Comment:
Maybe you can use a method to produce a VTO for the change instead of repeat code